lib/async_hooks.js

```diff
@@ -207,6 +208,128 @@ class AsyncResource {
+const contexts = new Set();
+const asyncContextHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    // save new resource
+    idResourcePairing.set(asyncId, new WeakReference(resource));
+    const triggerResourceWeak = idResourcePairing.get(triggerAsyncId);
+    contexts.forEach((context) => {
+      context._propagate(resource, triggerResourceWeak && triggerResourceWeak.get(), triggerAsyncId);
+    });
+  },
+  destroy(asyncId) {
+    idResourcePairing.delete(asyncId);
+  },
+  promiseResolve(asyncId) {
+    idResourcePairing.delete(asyncId);
+  }
+});
+
+class AsyncContext {
+
+  constructor() {
+    // WeakMap resource -> store
+    this._resourceStore = new WeakMap();
+    this.count = 0;
+    this._firstCalls = new Map();
+  }
+
+  /**
+   * propagate the context from a parent resource to a child one
+   */
+  _propagate(resource, triggerResource, triggerAsyncId) {
+    let store = this._resourceStore.get(triggerResource || {});
+    if (!store) {
+      const record = this._firstCalls.get(triggerAsyncId);
+      if (record instanceof Map) {
+        store = record;
+        // let's weaken the content of _firstCalls.get(triggerAsyncId)
+        // this is to prevent a potential memory leak
+        this._firstCalls.set(triggerAsyncId, new WeakReference(store));
+      }
+      if (record instanceof WeakReference) {
+        store = record.get();
+      }
+    }
+    if (store) {
+      this._resourceStore.set(resource, store);
+    }
+  }
+
+  /**
+   * get the current store, i.e. the unique map associated with current
+   * async resource and (maybe) inherited from the parent async resource
+   * will return null if no map is currently associated
+   */
+  getStore() {
+    const asyncId = executionAsyncId();
+    const resourceWeak = idResourcePairing.get(asyncId);
+    if (resourceWeak !== undefined) {
+      const resource = resourceWeak.get();
+      return this._resourceStore.get(resource);
+    }
+    // maybe what we want it in the _firstCalls map
+    const record = this._firstCalls.get(asyncId);
+    if (record === undefined) {
+      return; // definitely not found anywhere
+    }
+    if (record instanceof Map) {
+      // it has not been weakend yet
+      return record;
+    }
+    return record.get();
+  }
+
+  /**
+   * creates a parent async resource and associates a context with it
+   * then the callback is called in that new async context
+   * at this point calling getStore will return a map
+   * if called within the callback chain of another start call a new map will
+   * replace the previous one.
+   */
+  enter() {
+    if (contexts.size === 0){
+      asyncContextHook.enable();
+    }
+    contexts.add(this);
+    this.count++;
+    const store = new Map();
+    const asyncId = executionAsyncId();
+    const resourceWeak = idResourcePairing.get(asyncId);
+    if (resourceWeak !== undefined) {
+      const resource = resourceWeak.get();
+      this._resourceStore.set(resource, store);
```

Comment:
Could it happen that `resource` is `undefined` here which is invalid as key in a `WeakMap`?

Reply:
Good question. Indeed I don't think it could be `undefined`